### PR TITLE
Refactor interface items

### PR DESCRIPTION
**Pull request by osheamonn**
## Proposed changes

Refactors the InterfaceItem infrastructure to make better use of base tags, so that decisions about how quantities are computed on the interface are made when adding the Tag to the DataBox (is this a SimpleTag, a ComputeTag, or does it slice variables from the volume?) rather than at the level of the Tags inside of variables which is done now. 

Most of the significant changes occur in src/Domain/Tags.hpp. 
- InterfaceBase and InterfaceImpl removed. 
- InterfaceComputeItem and Slice added.
- Interface is now a simple item, which can be used as a base tag for the above two items. 

Breaks: 
- Anything that touches the Interface part of the Initialization code, the user must now specify 
whether Interface tags are simple, compute items or slices. Before, using the tag Interface worked for everything. 
- If `should_be_sliced_to_boundary` was being defined in any Tags, it now doesn't do anything. Whether the Tag should be sliced is specified by adding it to the DataBox using `Slice`. 

There are some TODO's scattered around which are questions I have for people, regarding using `tmpl::contains_v<TagsList, Tag>` to determine whether a Tag is in the DataBox, since it now fails for the compute tags or slice tags on the interface. 

The last commit (remove sliced_to...) should probably be squashed into the previous one, but I thought it would make reviewing easier.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Domain/Tags.hpp [hunk 1]
```diff
@@ -155,50 +155,39 @@ struct BoundaryDirections : db::ComputeTag {
...
-/// compute item's inputs should be taken from the volume even when
-/// applied on a slice, it may indicate them using a `volume_tags`
-/// typelist in the tag struct.
+/// The contained object will be a map from ::Direction to the item type of
+/// `Tag`, with the set of directions being those produced by `DirectionsTag`.
+///
+/// If a SimpleTag is desired on the interface, then this tag can be added to
+/// the DataBox directly. If a ComputeTag which acts on Tags on the interface is
+/// desired, then the tag should be added using `InterfaceComputeItem`. If a
+/// ComputeTag which slices a VariablesTag in the volume to an Interface, then
```
Comment (kidder) [inline src/Domain/Tags.hpp]: Interface is desired,

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_InitializeElement.cpp [hunk 2]
```diff
@@ -144,6 +144,15 @@ bool box_contains(const Box& /*box*/) noexcept {
   return tmpl::list_contains_v<typename Box::tags_list, Tag>;
 }
 
+// TODO: Because now (almost) all Interface tags are retrieved by base tags,
```
Comment (kidder) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_InitializeElement.cpp]: want to keep a runtime check

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by osheamonn** [comment]
Rebased and squashed. `tag_is_retrievable` is now done using a type trait which is checked at runtime.

**Review by kidder — DISMISSED** [review]

**Comment by osheamonn** [comment]
Had to rebase due to IWYU @kidder

**Review by kidder — APPROVED** [review]

**Review by mar-celine — APPROVED** [review]

**Comment by nilsdeppe** [comment]
travis failures are test timeouts. @wthrowe feel free to merge if you're happy with this :)

**Review by wthrowe — APPROVED** [review]

**Pull request merged**